# Assignment4 Week4

# Employee attrition


In [34]:
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import math

from sklearn.metrics import r2_score
import sklearn as sk
from sklearn.model_selection import train_test_split 

from sklearn.preprocessing import normalize
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn.tree import export_graphviz


In [35]:
df = pd.read_csv('Employee.csv')
df.head(50)

Age Attrition     BusinessTravel  DailyRate              Department  \
0    41       Yes      Travel_Rarely       1102                   Sales   
1    49        No  Travel_Frequently        279  Research & Development   
2    37       Yes      Travel_Rarely       1373  Research & Development   
3    33        No  Travel_Frequently       1392  Research & Development   
4    27        No      Travel_Rarely        591  Research & Development   
5    32        No  Travel_Frequently       1005  Research & Development   
6    59        No      Travel_Rarely       1324  Research & Development   
7    30        No      Travel_Rarely       1358  Research & Development   
8    38        No  Travel_Frequently        216  Research & Development   
9    36        No      Travel_Rarely       1299  Research & Development   
10   35        No      Travel_Rarely        809  Research & Development   
11   29        No      Travel_Rarely        153  Research & Development   
12   31        No      Travel_Rarely        670  Research & Development   
13   34        No      Travel_Rarely       1346  Research & Development   
14   28       Yes      Travel_Rarely        103  Research & Development   
15   29        No      Travel_Rarely       1389  Research & Development   
16   32        No      Travel_Rarely        334  Research & Development   
17   22        No         Non-Travel       1123  Research & Development   
18   53        No      Travel_Rarely       1219                   Sales   
19   38        No      Travel_Rarely        371  Research & Development   
20   24        No         Non-Travel        673  Research & Development   
21   36       Yes      Travel_Rarely       1218                   Sales   
22   34        No      Travel_Rarely        419  Research & Development   
23   21        No      Travel_Rarely        391  Research & Development   
24   34       Yes      Travel_Rarely        699  Research & Development   
25   53        No      Travel_Rarely       1282  Research & Development   
26   32       Yes  Travel_Frequently       1125  Research & Development   
27   42        No      Travel_Rarely        691                   Sales   
28   44        No      Travel_Rarely        477  Research & Development   
29   46        No      Travel_Rarely        705                   Sales   
30   33        No      Travel_Rarely        924  Research & Development   
31   44        No      Travel_Rarely       1459  Research & Development   
32   30        No      Travel_Rarely        125  Research & Development   
33   39       Yes      Travel_Rarely        895                   Sales   
34   24       Yes      Travel_Rarely        813  Research & Development   
35   43        No      Travel_Rarely       1273  Research & Development   
36   50       Yes      Travel_Rarely        869                   Sales   
37   35        No      Travel_Rarely        890                   Sales   
38   36        No      Travel_Rarely        852  Research & Development   
39   33        No  Travel_Frequently       1141                   Sales   
40   35        No      Travel_Rarely        464  Research & Development   
41   27        No      Travel_Rarely       1240  Research & Development   
42   26       Yes      Travel_Rarely       1357  Research & Development   
43   27        No  Travel_Frequently        994                   Sales   
44   30        No  Travel_Frequently        721  Research & Development   
45   41       Yes      Travel_Rarely       1360  Research & Development   
46   34        No         Non-Travel       1065                   Sales   
47   37        No      Travel_Rarely        408  Research & Development   
48   46        No  Travel_Frequently       1211                   Sales   
49   35        No      Travel_Rarely       1229  Research & Development   

    DistanceFromHome  Education    EducationField  EmployeeCount  \
0                  1          2     Life Sciences              1   
1                  8          1     Life S

In [36]:
dummies = pd.get_dummies(df['Gender'])
dummies.head()

Female  Male
0       1     0
1       0     1
2       0     1
3       1     0
4       0     1

In [37]:
df = pd.concat([df, dummies], axis=1) 
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0  ...                 0                 8                      0   
1  ...                 1                10                      3   
2  ...                 0                 7                      3   
3  ...                 0                 8                      3   
4  ...                 1                 6                      3   

   WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                1               6                  4   
1                3              10                  7   
2                3               0                  0   
3                3               8                  7   
4                3               2                  2   

   YearsSinceLastPromotion YearsWithCurrManager  Female  Male  
0                        0                    5       1     0  
1                        1                    7       0     1  
2                        0                    0       0     1  
3                        3                    0       1     0  
4                        2                    2       0     1  

[5 rows x 37 columns]

In [38]:
df_subset = df[['Age', 'DailyRate', 'Female', 'Attrition', 'Education', 'DistanceFromHome', 'TotalWorkingYears', 'YearsAtCompany']]
df_subset.head(10)

Age  DailyRate  Female Attrition  Education  DistanceFromHome  \
0   41       1102       1       Yes          2                 1   
1   49        279       0        No          1                 8   
2   37       1373       0       Yes          2                 2   
3   33       1392       1        No          4                 3   
4   27        591       0        No          1                 2   
5   32       1005       0        No          2                 2   
6   59       1324       1        No          3                 3   
7   30       1358       0        No          1                24   
8   38        216       0        No          3                23   
9   36       1299       0        No          3                27   

   TotalWorkingYears  YearsAtCompany  
0                  8               6  
1                 10              10  
2                  7               0  
3                  8               8  
4                  6               2  
5                  8               7  
6                 12               1  
7                  1               1  
8                 10               9  
9                 17               7

In [39]:
y = df_subset['Attrition'] #the indepent variable 
X = df_subset[['Age', 'DailyRate', 'Female', 'Education', 'DistanceFromHome', 'TotalWorkingYears', 'YearsAtCompany']] #the depentent variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #30- test/70- training relation  

In [40]:
model = RandomForestClassifier(max_depth=5, random_state=1)
model = model.fit(X, y)
y_test_pred = model.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix" on the test set
cm

/Users/quintennaayen/opt/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


array([[363,   1],
       [ 66,  11]])

In [41]:
conf_matrix = pd.DataFrame(cm, index=['NoAttrition_real', 'YesAttrition_real'], columns = ['NoAttrition_predict', 'YesAttrition_predict']) 
conf_matrix

NoAttrition_predict  YesAttrition_predict
NoAttrition_real                   363                     1
YesAttrition_real                   66                    11

The employer's confusion matrix dataset results in the following results: 
- 363 times it was well predicted that an employer had dropped out. only once was this prediction incorrect. 
- 66 times it was incorrectly predicted No attrition in stead of Yes Attrition.






In [42]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_test_pred))

              precision    recall  f1-score   support

          No       0.85      1.00      0.92       364
         Yes       0.92      0.14      0.25        77

    accuracy                           0.85       441
   macro avg       0.88      0.57      0.58       441
weighted avg       0.86      0.85      0.80       441



**Accuracy**

Accuracy = Correct/Total * 100 
374/441 * 100 = 84,8 %

**Precision** 

Precision talks about how precise/accurate your model is out of those predicted Attrition, how many of them are actual well predicted.

**Recall**

So Recall actually calculates how many of the Actual Yes our model capture through labeling it as Yes. Applying the same understanding, we know that Recall shall be the model metric we use to select our best model when there is a high cost associated with False No.

No Attrition:
precision = .85
recall = 1.00

Yes Attrition:
precision = 0.92
recall = 0.14